watchmal/engine/reconstruction.py

```diff
@@ -0,0 +1,422 @@
+            # save state at end of epoch
+            if self.rank == 0 and (save_interval is not None) and ((self.epoch+1) % save_interval == 0):
+                self.save_state(suffix=f'_epoch_{self.epoch+1}')
+        self.train_log.close()
+        if self.rank == 0:
+            log.info(f"Epoch {self.epoch} completed in {datetime.now() - epoch_start_time}")
+            log.info(f"Training {epochs} epochs completed in {datetime.now()-start_time}")
+            self.val_log.close()
+
+    def validate(self, val_iter, num_val_batches, checkpointing):
+        """
+        Perform validation with the current state, on a number of batches of the validation set.
+
+        Parameters
+        ----------
+        val_iter : iter
+            Iterator of the validation dataset.
+        num_val_batches : int
+            Number of validation batches to iterate over.
+        checkpointing : bool
+            Whether to save the current state to disk.
+        """
+        # set model to eval mode
+        self.model.eval()
+        val_metrics = None
+        for val_batch in range(num_val_batches):
+            # get validation data mini-batch
+            try:
+                val_data = next(val_iter)
+            except StopIteration:
+                del val_iter
+                if self.is_distributed:
+                    self.data_loaders["validation"].sampler.set_epoch(self.data_loaders["validation"].sampler.epoch+1)
+                val_iter = iter(self.data_loaders["validation"])
+                val_data = next(val_iter)
+            # extract the event data and target from the input data dict
+            self.data = val_data['data'].to(self.device)
+            self.target = val_data[self.truth_key].to(self.device)
+            # evaluate the network
+            outputs, metrics = self.forward(False)
+            if val_metrics is None:
+                val_metrics = metrics
+            else:
+                for k, v in metrics.items():
+                    val_metrics[k] += v
+        # record the validation stats to the csv
+        val_metrics = {k: v/num_val_batches for k, v in val_metrics.items()}
+        val_metrics = self.get_synchronized_metrics(val_metrics)
+        if self.rank == 0:
+            log_entries = {"iteration": self.iteration, "epoch": self.epoch, **val_metrics, "saved_best": False}
+            # Save if this is the best model so far
+            print(f"  Validation {', '.join(f'{k}: {v:.5g}' for k, v in val_metrics.items())}", end="")
+            if val_metrics["loss"] < self.best_validation_loss:
+                print(" ... Best validation loss so far!")
+                self.best_validation_loss = val_metrics["loss"]
+                self.save_state(suffix="_BEST")
+                log_entries["saved_best"] = True
+            else:
+                print("")
+            # Save the latest model if checkpointing
+            if checkpointing:
```

Comment:
Is there only checkpointing every epoch now? Previous version would only save when validation loss was lower.